heuristic/delta_calculations.py

```diff
@@ -272,173 +174,75 @@ def hard_constraint_penalties(state):
-                    state.soft_vars["isolated_working_days"][e, i + 1]
-                    + state.soft_vars["isolated_off_days"][e, i + 1]
-                    for e in employees
-                    for i in range(len(days_in_week) - 2)
-                )
-                - sum(
-                    state.soft_vars["consecutive_days"][e, i]
-                    for e in employees
-                    for i in range(len(days_in_week) - L_C_D)
-                )
-                - sum(
-                    state.hard_vars["below_minimum_demand"][c, t]
-                    + state.hard_vars["above_maximum_demand"][c, t]
-                    for c in competencies
-                    for j in weeks
-                    for t in time_periods_in_week[j]
-                )
-                - 10
-                * sum(
-                    state.hard_vars["more_than_one_shift_per_day"][e, i]
-                    for e in employees
-                    for i in days_in_week
-                )
-                - 10
-                * sum(
-                    state.hard_vars["cover_multiple_demand_periods"][e, t]
-                    for e in employees
-                    for j in weeks
-                    for t in time_periods_in_week[j]
-                )
-                - 10
-                * max(
-                    0, sum(state.soft_vars["deviation_contracted_hours"][e, j] for e in employees)
-                )
-                - 10 * sum(state.hard_vars["weekly_off_shift_error"][e, j] for e in employees)
-                - 100
-                * sum(state.hard_vars["delta_positive_contracted_hours"][e] for e in employees)
-            )
-    if setting == "worst":
+                - sum(state.soft_vars["isolated_working_days"][e, i + 1] + state.soft_vars["isolated_off_days"][e, i + 1] for e in employees for i in range(len(days_in_week)-2))
+                - sum(state.soft_vars["consecutive_days"][e,i] for e in employees for i in range(len(days_in_week)-L_C_D))
+                - 5 * sum(state.hard_vars["below_minimum_demand"][c, t] + state.hard_vars["above_maximum_demand"][c, t] for c in competencies for j in weeks for t in time_periods_in_week[c, j])
+                - 10 * sum(state.hard_vars["more_than_one_shift_per_day"][e, i] for e in employees for i in days_in_week)
+                - 10 * sum(state.hard_vars["cover_multiple_demand_periods"][e,t] for e in employees for j in weeks for t in combined_time_periods_in_week[j])
+                - 5 * max(0, sum(state.soft_vars["contracted_hours"][e,j] for e in employees))
+                - 10 * sum(state.hard_vars["weekly_off_shift_error"][e,j] for e in employees)
+                - 100 * sum(state.hard_vars["delta_positive_contracted_hours"][e] for e in employees)
+                - 100 * competency_score
+                )
+    if(setting == "worst"):
         value = sorted(value, key=value.get, reverse=False)[:k]
         return value
     else:
         return list(value.values())
```

Comment:
Why are you converting to a list here? I think we should consider using list all over the place for f as it is much faster to sum over a list then a dict, altough lookup could be worse for lager number

Reply:
Actually, on second thought, I think a HeapQueue of Tuples could be really beneficial. It is clearly more work to implement, but we can practically eliminate the cost of sorting with a heap, making it really useful for remove-worst operators.

https://stackoverflow.com/questions/45137002/define-heap-key-for-an-array-of-tuples/45137076

Reply:
For initial tests a heap leaves a list in the dust: 

<img width="434" alt="bilde" src="https://user-images.githubusercontent.com/1371159/81442121-4ac65580-9173-11ea-9c6a-a752ab9458ef.png">

Reply:
I see now that your method acutally utilizes heaps, so this comment is not relevant in this use case, but there might be others. 

https://docs.python.org/3/library/heapq.html#heapq.nsmallest